### PR TITLE
[MSFT] Factor out Quartus/Stratix10 logic from TclEmitter

### PR DESCRIPTION
**Pull request by mortbopet**
Refactoring commit which makes `TclEmitter` generic by allowing inheriting classes to provide custom `TclOutputState` classes. Furthermore, adds new arguments to `msft-export-tcl` to select tcl flavor and target device, respectively.
The intention of this refactor is to facilitate writing TCL emitters for alternative flows and devices.

### CODE REVIEW

File: include/circt/Dialect/MSFT/TclEmitter.h [hunk 1]
```diff
@@ -48,7 +93,9 @@ class TclEmitter {
       tclOpsForModInstance;
   DenseSet<hw::GlobalRefOp> refsUsed;
 
+  // Populates the symbol cache.
   LogicalResult populate();
+  bool populated = false;
```
Comment (teqdruid) [inline include/circt/Dialect/MSFT/TclEmitter.h:98]: Setting `false` here is redundant with the constructor.

Comment (teqdruid) [reply include/circt/Dialect/MSFT/TclEmitter.h:98]: Can you move `topLevelSymbols` down as well. I'd like to keep these grouped.

**Review by teqdruid — CHANGES_REQUESTED** [review]
The Quartus API won't (or is very unlikely to) change radically between families. If anything, it could change version-to-version! So please drop the S10 stuff. If we find it necessary to make small tweaks on a per-family basis (e.g. the DSPs in other families aren't called MPDSP), we should add special cases since 99% of the output is the same.

File: include/circt/Dialect/MSFT/TclEmitter.h [hunk 2]
```diff
@@ -14,8 +14,11 @@
 #define CIRCT_DIALECT_MSFT_EXPORTTCL_H
 
 #include "circt/Dialect/HW/HWSymCache.h"
+#include "circt/Dialect/MSFT/MSFTAttributes.h"
 #include "circt/Dialect/MSFT/MSFTOpInterfaces.h"
+#include "circt/Dialect/MSFT/MSFTOps.h"
 #include "circt/Support/LLVM.h"
+#include "llvm/Support/raw_ostream.h"
```
[the remark this answers is not in the document]
Comment (mortbopet) [reply include/circt/Dialect/MSFT/TclEmitter.h:22]: To me, a rename makes sense to reflect the class that's implemented in the file through the filename (`TclEmitter.h` implements `class TclEmitter`) - even more so, now that other files are expected to subclass `TclEmitter`.

File: include/circt/Dialect/MSFT/MSFTPasses.td [hunk 3]
```diff
@@ -25,7 +25,11 @@ def ExportTcl: Pass<"msft-export-tcl", "mlir::ModuleOp"> {
                "List of top modules to export Tcl for",
                "llvm::cl::ZeroOrMore,">,
     Option<"tclFile", "tcl-file", "std::string",
-           "", "File to output Tcl into">
+           "", "File to output Tcl into">,
+    Option<"flavor", "flavor", "std::string",
+           "\"quartus\"", "Target TCL flavor. Options: [quartus]">,
+    Option<"device", "device", "std::string",
```
[the remark this answers is not in the document]
Comment (mortbopet) [reply include/circt/Dialect/MSFT/MSFTPasses.td:31]: Will drop device for now.

File: lib/Dialect/MSFT/MSFTPasses.cpp [hunk 4]
```diff
@@ -474,19 +475,26 @@ struct ExportTclPass : public ExportTclBase<ExportTclPass> {
 void ExportTclPass::runOnOperation() {
   auto top = getOperation();
   auto *ctxt = &getContext();
-  TclEmitter emitter(top);
+  std::unique_ptr<TclEmitter> emitter;
+  if (tool == "quartus")
```
Comment (teqdruid) [inline lib/Dialect/MSFT/MSFTPasses.cpp:479]: Please surround both blocks with curly brackets if one is necessary.

**Review by teqdruid — APPROVED** [review]
lgtm. thanks!

**Pull request closed without merging**